nav2_costmap_2d/plugins/range_sensor_layer.cpp

```diff
@@ -0,0 +1,487 @@
+{
+  double lbda = delta(phi) * gamma(theta);
+
+  double delta = resolution_;
+
+  if (phi >= 0.0 && phi < r - 2 * delta * r) {
+    return (1 - lbda) * (0.5);
+  } else if (phi < r - delta * r) {
+    return lbda * 0.5 * pow((phi - (r - 2 * delta * r)) / (delta * r), 2) +
+           (1 - lbda) * .5;
+  } else if (phi < r + delta * r) {
+    double J = (r - phi) / (delta * r);
+    return lbda * ((1 - (0.5) * pow(J, 2)) - 0.5) + 0.5;
+  } else {
+    return 0.5;
+  }
+}
+
+void RangeSensorLayer::bufferIncomingRangeMsg(
+  const sensor_msgs::msg::Range::SharedPtr range_message)
+{
+  range_message_mutex_.lock();
+  range_msgs_buffer_.push_back(*range_message);
+  range_message_mutex_.unlock();
+}
+
+void RangeSensorLayer::updateCostmap()
+{
+  std::list<sensor_msgs::msg::Range> range_msgs_buffer_copy;
+
+  range_message_mutex_.lock();
+  range_msgs_buffer_copy = std::list<sensor_msgs::msg::Range>(range_msgs_buffer_);
+  range_msgs_buffer_.clear();
+  range_message_mutex_.unlock();
+
+  for (auto & range_msgs_it : range_msgs_buffer_copy) {
+    processRangeMessageFunc_(range_msgs_it);
+  }
+}
+
+void RangeSensorLayer::processRangeMsg(sensor_msgs::msg::Range & range_message)
+{
+  if (range_message.min_range == range_message.max_range) {
+    processFixedRangeMsg(range_message);
+  } else {
+    processVariableRangeMsg(range_message);
+  }
+}
+
+void RangeSensorLayer::processFixedRangeMsg(sensor_msgs::msg::Range & range_message)
+{
+  if (!std::isinf(range_message.range)) {
+    RCLCPP_ERROR(
+      node_->get_logger(),
+      "Fixed distance ranger (min_range == max_range) in frame %s sent invalid value. "
+      "Only -Inf (== object detected) and Inf (== no object detected) are valid.",
+      range_message.header.frame_id.c_str());
+    return;
+  }
+
+  bool clear_sensor_cone = false;
+
+  if (range_message.range > 0) {  // +inf
+    if (!clear_on_max_reading_) {
+      return;  // no clearing at all
+    }
+    clear_sensor_cone = true;
+  }
+
+  range_message.range = range_message.min_range;
+
+  updateCostmap(range_message, clear_sensor_cone);
+}
+
+void RangeSensorLayer::processVariableRangeMsg(sensor_msgs::msg::Range & range_message)
+{
+  if (range_message.range < range_message.min_range || range_message.range >
+    range_message.max_range)
+  {
+    return;
+  }
+
+  bool clear_sensor_cone = false;
+
+  if (range_message.range == range_message.max_range && clear_on_max_reading_) {
+    clear_sensor_cone = true;
+  }
+
+  updateCostmap(range_message, clear_sensor_cone);
+}
+
+void RangeSensorLayer::updateCostmap(
+  sensor_msgs::msg::Range & range_message,
+  bool clear_sensor_cone)
+{
+  max_angle_ = range_message.field_of_view / 2;
+
+  geometry_msgs::msg::PointStamped in, out;
+  in.header.stamp = range_message.header.stamp;
+  in.header.frame_id = range_message.header.frame_id;
+
+  if (!tf_->canTransform(
+      global_frame_, in.header.frame_id, in.header.stamp,
+      rclcpp::Duration(1e8)))
```

Comment:
1e8?

Reply:
Do you have a suggestion for a different timeout value or are you looking for a chrono literal to denote the units?

Reply:
@Michael-Equi what do other layers do for the transform timeouts? Do that for consistency.